public_html/wp-content/mu-plugins/utilities/class-meetup-oauth2-client.php

```diff
@@ -0,0 +1,460 @@
+			 * error. In practice we've observed that it's common for a seemingly valid request to be rejected with
+			 * a `400` response, but then get a `200` response if that exact same request is retried.
+			 */
+			'breaking_response_codes' => array(
+				401, // Unauthorized (invalid key).
+				429, // Too many requests (rate-limited).
+				404, // Unable to find group
+			),
+		) );
+	}
+
+	/**
+	 * Generate the headers for a request.
+	 *
+	 * @param string $access_token Optional. Providing an access token will add an extra header.
+	 *
+	 * @return array
+	 */
+	protected function get_headers( $access_token = '' ) {
+		$headers = array(
+			'Accept' => 'application/json',
+		);
+
+		if ( $access_token ) {
+			$headers['Authorization'] = "Bearer $access_token";
+		}
+
+		return $headers;
+	}
+
+	/**
+	 * Step 1 in "Server Flow with User Credentials".
+	 *
+	 * @see https://www.meetup.com/meetup_api/auth/#oauth2servercredentials-auth
+	 *
+	 * @return string
+	 */
+	protected function request_authorization_code() {
+		$authorization_code = '';
+
+		$request = array(
+			'client_id'     => self::CONSUMER_KEY,
+			'redirect_uri'  => self::REDIRECT_URI,
+			'response_type' => 'anonymous_code',
+			'scope'         => 'ageless',
+		);
+
+		$args = array(
+			'headers' => $this->get_headers(),
+			'body'    => $request,
+		);
+
+		$response = $this->tenacious_remote_post( self::URL_AUTHORIZE, $args );
+
+		if ( 200 === wp_remote_retrieve_response_code( $response ) ) {
+			$body = json_decode( wp_remote_retrieve_body( $response ), true );
+
+			if ( ! empty( $body['code'] ) ) {
+				$authorization_code = $body['code'];
+			} else {
+				$this->error->add(
+					'unexpected_oauth_authorization_code_response',
+					'The Meetup OAuth API response did not provide the expected data.'
+				);
+			}
+		} else {
+			$this->handle_error_response( $response );
+		}
+
+		return $authorization_code;
+	}
+
+	/**
+	 * Step 2 in "Server Flow with User Credentials".
+	 *
+	 * @see https://www.meetup.com/meetup_api/auth/#oauth2servercredentials-access
+	 *
+	 * Also the step for refreshing an expired access token.
+	 *
+	 * @see https://www.meetup.com/meetup_api/auth/#oauth2server-refresh
+	 *
+	 * Note that this does not store the access token. See get_access_token().
+	 *
+	 * @param string $type The type of grant. Either 'anonymous_code' or 'refresh_token'.
+	 * @param string $code The code/token used with the grant.
+	 *
+	 * @return array A successfully retrieved token will be an associative array with several keys.
+	 */
+	protected function request_access_token( $type, $code ) {
+		$access_token = array();
+
+		$request = array(
+			'client_id'     => self::CONSUMER_KEY,
+			'client_secret' => self::CONSUMER_SECRET,
+			'redirect_uri'  => self::REDIRECT_URI,
+			'grant_type'    => $type,
+		);
+
+		switch( $type ) {
+			case 'anonymous_code':
+				$request['code'] = $code;
+				break;
+			default:
+				$request[ $type ] = $code;
+				break;
+		}
```

Comment:
`SHOULD` - It's not obvious to me what's going on here. Is one of these for getting a new code, and the other for renewing an expired code? It'd be helpful to add inline comments for each saying what they're used for.

Reply:
Fixed in 0f24fca, but then I ended up doing a significant refactor in ea24b48. Might be worth taking another look to see if it's understandable now.

Reply:
Yeah, this looks good. Kudos for persevering through their inaccurate docs!